flowmachine/flowmachine/features/subscriber/meaningful_locations.py

```diff
@@ -0,0 +1,259 @@
+        GROUP BY subscriber, label, cluster
+        ORDER BY subscriber
+        """
+
+
+class MeaningfulLocationsAggregate(Query):
+    """
+    Aggregates an individual level meaningful location to a spatial unit by assigning
+    subscribers with clusters in that unit to it. For subscribers with more than one cluster,
+    assigns `1/n_clusters` to each spatial unit that the cluster lies in.
+
+    Parameters
+    ----------
+    meaningful_locations : MeaningfulLocations
+        A per-subscriber meaningful locations object to aggregate
+    level : {"admin3", "admin2", "admin1", "grid", "polygon"}, default "admin3"
+        Spatial unit to aggregate to
+    column_name : str or list of str, default None
+        Optionally specify a non-default column name or names from the spatial unit table
+    polygon_table : str, default None
+        When using the "polygon" level, you must specify the fully qualified name of a table
+        containing polygons.
+    geom_column : str, default "geom"
+        When using the "polygon" level, you must specify the name of column containing geometry
+    size : int, default None
+        When using the "grid" level, you must specify the size of the grid to use in KM
+    """
+
+    allowed_levels = {"admin3", "admin2", "admin1", "grid", "polygon"}
+
+    def __init__(
+        self,
+        *,
+        meaningful_locations: MeaningfulLocations,
+        level: str = "admin3",
+        column_name: Union[str, None, List[str]] = None,
+        polygon_table: str = None,
+        geom_column: str = "geom",
+        size: int = None,
+    ) -> None:
+        self.meaningful_locations = meaningful_locations
+        level_cols = get_columns_for_level(level, column_name)
+        self.column_name = column_name
+        self.level = level
+        if level.startswith("admin"):
+            if level_cols == ["pcod"]:
+                level_cols = [f"{level}pcod"]
+            self.aggregator = GeoTable(
+                f"geography.{level}", geom_column="geom", columns=["geom"] + level_cols
+            )
+        elif level == "polygon":
+            self.aggregator = GeoTable(
+                polygon_table,
+                geom_column=geom_column,
+                columns=[geom_column] + level_cols,
+            )
+        elif level == "grid":
+            self.aggregator = Grid(size=size)
+        else:
+            raise BadLevelError(
+                f"'{level}' is not an allowed level for meaningful locations, must be one of {MeaningfulLocationsOD.allowed_levels}'"
+            )
+        super().__init__()
+
+    @property
+    def column_names(self) -> List[str]:
+        return (
+            ["label"] + get_columns_for_level(self.level, self.column_name) + ["total"]
+        )
+
+    def _make_query(self):
+        agg_query, agg_cols = self.aggregator._geo_augmented_query()
+        level_cols = get_columns_for_level(self.level, self.column_name)
```

Comment:
Note to self: a _lot_ of this "levels" logic is duplicated and spread all over the place. Will be good to make this the focus of one of the next refactorings to encapsulate it (and esp. allow treating all aggregation units uniformly) so that no other code needs to know about and deal with this logic.